accepted/2022/capability-analyzer/capability-analyzer.md

```diff
@@ -0,0 +1,205 @@
+  no parameters or a property. In either case the return type must be of type
+  `bool`. A value of `true` indicates that the capability exists. The
+  `[CapabilityCheck]` attribute can be applied multiple times, which would
+  indicate different ways the caller can check.
+* In addition, a consumer can introduce their own capability checks by marking
+  their APIs with the `[CapabilityGuard]` attribute. They reference the type of
+  capability attribute the guard is asserting. The guard can either be a method,
+  property, or a field. In case of methods, parameters are also permissible.
+  However, the return value must of be of type `bool`. A value of `true` implies
+  the capability is there. Custom guards are useful for cases where a feature
+  requires multiple checks. We recently added this to our [platform
+  compatibility analyzer][platform-compat] and it has proven to be invaluable
+  because it makes it much easier to enable code analysis for existing code
+  bases where capabilities are already checked as part of the infrastructure.
+
+### Threading Annotations
+
+In order to model the thread creation annotation, we need an attribute for the
+capability and a new capability API on `Thread`:
+
+```C#
+namespace System.Threading
+{
+    [AttributeUsage(AttributeTargets.All, AllowMultiple = false, Inherited = false)]
+    [CapabilityCheck(typeof(Thread), nameof(Thread.IsCreationSupported))]
+    public sealed class RequiresThreadCreationAttribute : CapabilityAttribute
+    {
+        public RequiresThreadCreationAttribute();
+    }
+
+    public partial class Thread
+    {
+        public static bool IsCreationSupported { get; }
+
+        [RequiresThreadCreation]
+        public void Start();
+    }
+}
+```
+
+### Reflection Emit
+
+In order to model dynamic code generation we can probably reuse the existing
+linker annotation `RequiresDynamicCodeAttribute` and change the attribute to
+inherit from `CapabilityAttribute`. For the capability check we can reuse the
+existing `IsDynamicCodeSupported` and `IsDynamicCodeCompiled` members on
+`RuntimeFeature`.
+
+**OPEN QUESTION** Can we reuse the `RequiresDynamicCodeAttribute`? Currently
+that's only used by the linker, but the semantics would be equivalent.
+
+```C#
+namespace System.Reflection.Emit
+{
+    [CapabilityCheck(typeof(RuntimeFeature), nameof(RuntimeFeature.IsDynamicCodeSupported))]
+    [CapabilityCheck(typeof(RuntimeFeature), nameof(RuntimeFeature.IsDynamicCodeCompiled))]
+    public sealed class RequiresDynamicCodeAttribute : CapabilityAttribute
+    {
+        public RequiresDynamicCodeAttribute();
+    }
+
+    [RequiresDynamicCode]
+    public partial class DynamicMethod : MethodInfo
+    {
+    }
+
+    [RequiresDynamicCode]
+    public partial class AssemblyBuilder : Assembly
+    {
+    }
+
+    // Would also mark
+    // - Other *Builder types
+    // - AppDomain.DefineDynamicAssembly
+    // - ILGenerator
+}
+```
+
```

Comment:
We also have `RequiresAssemblyFilesAttribute` and `RequiresUnreferencedCodeAttribute` and associates analyzers. Are these attributes in scope or out of scope for this proposal?

Reply:
Depends on whether what they track can be mapped to the capability analyzer. The goal wasn't to replace the existing linker analyzers as much as it was to replace the `[UnsupportedOSPlatform("browser")]` annotations with something more scoped/future proof. I merely added the dynamic code capability because it was a recent example of a capability-based API that this design *should* be able to accommodate as well.